File: py/0648-replace-words.py
from collections import defaultdict


class TrieNode:
    def __init__(self) -> None:
        self.children: defaultdict[str, TrieNode] = defaultdict(TrieNode)
        self.is_end = False


class Solution:
    def replaceWords(self, dictionary: list[str], sentence: str) -> str:
        root = TrieNode()
        for word in dictionary:
            node = root
            for c in word:
                node = node.children[c]
            node.is_end = True

        def replace(word: str) -> str:
            node = root
            for i, c in enumerate(word):
                if node.is_end:
                    return word[:i]
                node = node.children[c]
            return word

        return " ".join(replace(word) for word in sentence.split())

File: cpp/0648-replace-words.cpp
#include <sstream>
#include <string>
#include <unordered_map>
#include <vector>

class TrieNode {
    public:
        std::array<TrieNode *, 26> children{};
        bool is_end{false};
};

class Solution {
    public:
        [[nodiscard]] auto replaceWords(
            const std::vector<std::string> &dictionary,
            const std::string &sentence
        ) -> std::string {
            auto root = TrieNode{};
            for (const auto &word : dictionary) {
                auto node = &root;
                for (const auto c : word) {
                    const auto i = static_cast<unsigned>(c - 'a');
                    if (node->children[i] == nullptr) {
                        node->children[i] = new TrieNode{};
                    }
                    node = node->children[i];
                }
                node->is_end = true;
            }

            auto ss = std::istringstream{sentence};
            auto word = std::string{};
            word.reserve(1000);
            auto prefix = std::string{};
            prefix.reserve(100);
            auto ans = std::string{};
            ans.reserve(sentence.size() + 1);
            while (ss >> word) {
                const auto *node = &root;
                for (const auto c : word) {
                    const auto i = static_cast<unsigned>(c - 'a');
                    const auto *const child = node->children[i];
                    if (child == nullptr) {
                        break;
                    }
                    prefix += c;
                    node = child;
                    if (node->is_end) {
                        break;
                    }
                }
                ans += node->is_end ? prefix : word;
                ans += ' ';
                prefix.clear();
            }

            // remove last space
            ans.pop_back();

            return ans;
        }
};
